File: client/typescript/src/utils.test.ts
import { expect, it } from "vitest";
import { allowSandbox } from "./utils";
import type { EgressRule, SandboxConfig } from "./schemas";

function pinnedBody(rules: EgressRule[]): SandboxConfig {
  const create = rules.find((r) => r.override?.body_strategy === "replace");
  expect(create).toBeDefined();
  return create!.override!.body as SandboxConfig;
}

// The pinned body replaces the nested create's request body verbatim, skipping
// getOrCreateSandbox's client-side defaulting. Without the same defaults a
// config with no `fs` creates a sandbox with NO workspace — while a resumed VM
// snapshot still holds its 9p /workspace mount, which is then orphaned and
// wedges the first process to touch it in p9_client_rpc.
it("allowSandbox pins the default /workspace fs when config has none", () => {
  const body = pinnedBody(
    allowSandbox("worker-1", {
      image: "browser",
      snapshot: { vm: { key: "browser" } },
    }),
  );
  expect(body.fs).toEqual([
    {
      backend: "local",
      mount: "/workspace",
      acls: [{ path: "/workspace/**", access: "rw" }],
    },
  ]);
  expect(body.egress).toEqual([{ host: "*", access: "allow" }]);
  expect(body.image).toBe("browser");
  expect(body.snapshot).toEqual({ vm: { key: "browser" } });
});

it("allowSandbox keeps an explicit fs and egress unchanged", () => {
  const fs = [{ backend: "local" as const, mount: "/data" }];
  const egress = [{ host: "example.com", access: "allow" as const }];
  const body = pinnedBody(allowSandbox("worker-1", { fs, egress }));
  expect(body.fs).toEqual(fs);
  expect(body.egress).toEqual(egress);
});

it("allowSandbox pins the same body on every create rule", () => {
  const rules = allowSandbox("worker-1", { image: "browser" });
  const bodies = rules
    .filter((r) => r.override?.body_strategy === "replace")
    .map((r) => r.override!.body);
  expect(bodies).toHaveLength(2); // docker + k8s gateway hosts
  expect(bodies[0]).toEqual(bodies[1]);
});

File: client/go/utils_test.go
package client

import "testing"

// The pinned body replaces the nested create's request body verbatim, skipping
// GetOrCreateSandbox's client-side defaulting. Without the same defaults a
// config with no FS creates a sandbox with NO workspace — while a resumed VM
// snapshot still holds its 9p /workspace mount, which is then orphaned and
// wedges the first process to touch it in p9_client_rpc.
func TestAllowSandbox_PinsDefaultWorkspaceFS(t *testing.T) {
	rules := AllowSandbox("worker-1", SandboxConfig{Image: "browser"})

	var bodies []SandboxConfig
	for _, r := range rules {
		if r.Override != nil && r.Override.BodyStrategy == BodyStrategyReplace {
			cfg, ok := r.Override.Body.(SandboxConfig)
			if !ok {
				t.Fatalf("pinned body is %T, want SandboxConfig", r.Override.Body)
			}
			bodies = append(bodies, cfg)
		}
	}
	if len(bodies) != 2 { // docker + k8s gateway hosts
		t.Fatalf("got %d create rules with a pinned body, want 2", len(bodies))
	}
	for _, cfg := range bodies {
		if len(cfg.FS) != 1 || cfg.FS[0].Mount != "/workspace" || cfg.FS[0].Backend != "local" {
			t.Errorf("pinned FS = %+v, want the default /workspace local mount", cfg.FS)
		}
		if len(cfg.Egress) != 1 || cfg.Egress[0].Host != "*" || cfg.Egress[0].Access != "allow" {
			t.Errorf("pinned Egress = %+v, want the default allow-all rule", cfg.Egress)
		}
		if cfg.Image != "browser" {
			t.Errorf("pinned Image = %q, want %q", cfg.Image, "browser")
		}
	}
}

func TestAllowSandbox_KeepsExplicitFS(t *testing.T) {
	explicit := []FileSystem{{Mount: "/data", Backend: "local"}}
	rules := AllowSandbox("worker-1", SandboxConfig{FS: explicit})
	for _, r := range rules {
		if r.Override == nil {
			continue
		}
		cfg := r.Override.Body.(SandboxConfig)
		if len(cfg.FS) != 1 || cfg.FS[0].Mount != "/data" {
			t.Errorf("pinned FS = %+v, want the explicit /data mount", cfg.FS)
		}
	}
}
